New functions and some mods to old ones

kathryn-willi opened the pull request:
I've added two new functions:

1. `TADA_MakeSpatial()`: Per your previous recommendation, I created a new function that takes in a WQP data frame and converts it into an {sf} object.
2. `TADA_DataRetrieval_new()`: This will theoretically replace TADA's current `TADA_DataRetrieval()` function. It is identical to the original, except it now has an argument for pulling in WQP data based on a user-supplied shapefile.

I also modified the two functions you've already reviewed:

1. `TADA_GetATTAINS()`: I added some error handling, and I also added an argument to the function for the user to choose whether they want the ATTAINS shapefiles returned in their local environment (return = TRUE/FALSE).
2. `TADA_ViewATTAINS()`: It looks like I made a lot of changes, but really all I did was add a legend and changed the color scheme. 

**My key requests for this PR:**
- Like last time, I'd love it if you played around with each of these functions on new data pulls. Explore your hometown or something 😸. I am hopeful all errors are now handled properly! 
- I am also hoping to get some feedback on how I created the `return = T/F` argument in `TADA_GetATTAINS()` - is this a weird thing for a function to do? I thought about making the export of ATTAINS data its own function (see `scratch/TADA_ExportATTAINS()`), but for that, the `return()` object would be a list of nested shapefiles... this felt a bit clunky and potentially problematic for novice R users?
- Lastly, I am hoping to share these functions with the EPA ASAP. Do you think these functions are ready for their official review? I have tried my best to keep my code as similar to the TADA-style syntax as possible. If you agree they are in okay shape, my step will be to create some simple vignettes for each function - likely based on what I've done for the RMD in this repo. 

Thanks so much!
Katie

kathryn-willi committed:
need tigris

kathryn-willi committed:
ignore testing sf's etc.

kathryn-willi committed:
add sf argument to TADA_DataRetrieval

kathryn-willi committed:
add sf to current TADA function

kathryn-willi committed:
add legend + error handling

kathryn-willi committed:
new function to "spatialize" WQP data

kathryn-willi committed:
error handling + option to return ATTAINS objects

kathryn-willi committed:
scratch function (don't want to get rid of just yet!)

kathryn-willi committed:
updated

kathryn-willi committed:
updates

steeleb commented:
Heya! I'm pretty tied up until Thursday, but if there is anything small/specific I can do to help, let me know and I'll do it first thing tomorrow.

kathryn-willi commented:
> Heya! I'm pretty tied up until Thursday, but if there is anything small/specific I can do to help, let me know and I'll do it first thing tomorrow.

All good, B! @mbrousil can you take a look at this by the end of the week?

mbrousil commented:
> > Heya! I'm pretty tied up until Thursday, but if there is anything small/specific I can do to help, let me know and I'll do it first thing tomorrow.
> 
> All good, B! @mbrousil can you take a look at this by the end of the week?

Yep! I just opened it up locally and I'll work through it today and tomorrow.

mbrousil commented on new_tada_functions.Rmd:
````diff
@@ -9,74 +9,156 @@ output:
+spatial <- TADA_MakeSpatial(data = not_spatial, crs = 4269)
+```
+
+## `TADA_DataRetrieval()` - modified!
````
I'm not sure what the future of this doc is, but I was briefly confused because `_new()` wasn't included in the name, so you might consider clarifying.

mbrousil commented on new_tada_functions.Rmd:
```diff
@@ -9,74 +9,156 @@ output:
+                                           characteristicName = "pH",
+                                           sf = sf_example,
+                                           statecode = "AZ",
+                                           # only data within sf that is also in Washington County
```
```suggestion
                                           # only data within sf that is also in Maricopa County
```

mbrousil commented on src/TADA_DataRetrieval_new.R:
```diff
@@ -0,0 +1,369 @@
+#' @param countycode FIPS county name. Note that a state code must also be supplied (e.g. statecode = "AL", countycode = "Chilton").
+#' @param huc A numeric code denoting a hydrologic unit. Example: "04030202". Different size hucs can be entered.
+#' @param siteid Unique monitoring location identifier
+#' @param sf A polygon to subset the WQP data by
```
I wonder if an argument name like `shape` or `aoi` might be clearer than `sf`?

kathryn-willi replied:
i like that!

mbrousil commented on src/TADA_DataRetrieval_new.R:
```diff
@@ -0,0 +1,369 @@
+    }
+  }
+  
+  # subset original dataframe to only those WQP observations within the user-supplied sf:
+  if (!is.null(sf) & inherits(sf, "sf")) {
+    try(TADAprofile.clean <- TADAprofile.clean %>%
+          # transform df into spatial object using a new TADA_MakeSpatial() function
+          TADA_MakeSpatial(data = ., crs = 4326) %>%
+          # transform to the CRS of the sf object
+          sf::st_transform(sf::st_crs(sf)) %>%
+          # select only WQP observations within the sf object
+          .[sf, ] %>%
+          sf::st_drop_geometry(),
+        silent = TRUE)
+  }
```
Would it be possible (or useful?) to directly submit spatial data in the original query instead of using it to subset after downloading? That's what I was anticipating this function would do at least, but I'm not sure if it is the right thing to do. For example: https://github.com/DOI-USGS/ds-pipelines-targets-example-wqp/blob/e796fc7a792986865f6f77e184f18986073ca8ea/1_inventory/src/get_wqp_inventory.R#L61-L66

kathryn-willi replied:
I think that could definitely speed things up!

mbrousil commented on src/TADA_GetATTAINS.R:
```diff
@@ -75,21 +63,93 @@ TADA_GetATTAINS <- function(data){
+          dplyr::distinct(.keep_all = TRUE),
+        silent = TRUE)
+    # if not ATTAINS data return original data frame:
+    if(is.null(nearby_catchments)){
+      print("There are no ATTAINS features associated with these WQP observations. Returning original dataframe.")
+      return(data)
```
thought: Would it be better to have this produce an error? When reading this made me wonder if someone trying to iterate with this could inadvertently use the original dataset, which wouldn't have the same structure as what they'd anticipate as output.

I think that it would also help to mention the "no ATTAINS" behavior in the function documentation.

kathryn-willi replied:
I agree - this was something done in another TADA function that I copied. But your suggestion makes more sense! OR - I could add those ATTAINS columns and they are just empty...

mbrousil replied:
Ooooh yep that seems viable too!

mbrousil commented on src/TADA_GetATTAINS.R:
```diff
@@ -75,21 +63,93 @@ TADA_GetATTAINS <- function(data){
+      
+      if(return == TRUE){
+        
+        download_attains <- function(baseurl){
+          # bounding box (with some wiggle) of TADA data
+          bbox <- TADA_DataRetrieval_data %>% 
+            sf::st_buffer(0.001) %>% 
+            sf::st_bbox(TADA_DataRetrieval_data)
+          
+          # convert bounding box to characters
+          bbox <- toString(bbox) %>% 
+            # encode for use within the API URL
+            urltools::url_encode(.)
+          
+          # EPSG for CRS used by TADA data. THIS NEEDS TO BE IMPROVED!
+          epsg <- sf::st_crs(TADA_DataRetrieval_data)$epsg
+          
+          # set all necessary parameters for the query
+          query <- urltools::param_set(baseurl, key = "geometry", value = bbox) %>%
+            urltools::param_set(key = "inSR", value = epsg) %>%
+            urltools::param_set(key = "resultRecordCount", value = 5000) %>%
+            urltools::param_set(key = "spatialRel", value = "esriSpatialRelIntersects") %>%
+            urltools::param_set(key = "f", value = "geojson") %>%
+            urltools::param_set(key = "outFields", value = "*") %>%
+            urltools::param_set(key = "geometryType", value = "esriGeometryEnvelope") %>%
+            urltools::param_set(key = "returnGeometry", value = "true") %>%
+            urltools::param_set(key = "returnTrueCurves", value = "false") %>%
+            urltools::param_set(key = "returnIdsOnly", value = "false") %>%
+            urltools::param_set(key = "returnCountOnly", value = "false") %>%
+            urltools::param_set(key = "returnZ", value = "false") %>%
+            urltools::param_set(key = "returnM", value = "false") %>%
+            urltools::param_set(key = "returnDistinctValues", value = "false") %>%
+            urltools::param_set(key = "returnExtentOnly", value = "false") %>%
+            urltools::param_set(key = "featureEncoding", value = "esriDefault")
+          
+          final_sf <- geojsonsf::geojson_sf(query)
+          
+          return(final_sf)
+          
+        }
```
suggestion (non-blocking): If it doesn't affect the functionality of the code, I'd suggest moving this outside the main function definition but within the same script. Perhaps above this function? You could combine multiple functions that use it into a single script, or have it sourced, etc. It would make it easier to follow both functions defined in this script.

mbrousil commented on src/TADA_MakeSpatial.R:
```diff
@@ -0,0 +1,54 @@
+    # For now assume these are WGS84. USGS has done this, too. 
+    "UNKWN", 4326,
+    "OTHER", 4326,
+    # There's some others I've encountered too, e.g.
```
```suggestion
```

mbrousil commented on src/TADA_MakeSpatial.R:
```diff
@@ -0,0 +1,54 @@
+    "NAD83", 4269,
+    "WGS84", 4326,
+    "NAD27", 4267,
+    # For now assume these are WGS84. USGS has done this, too. 
```
```suggestion
    # Assume these are WGS84
```

mbrousil commented on src/TADA_GetATTAINS.R:
```diff
@@ -75,21 +63,93 @@ TADA_GetATTAINS <- function(data){
+          return(final_sf)
+          
+        }
+        
```
Is it possible to add more high-level explanatory comments here? I'm having trouble figuring out what's going on from here to the end of the script.

mbrousil reviewed (commented):
Hey Katie! I've added comments to some individual places in the files where I've got targeted questions, suggestions, etc. Below I'll also put some bigger picture thoughts. My opinion is that things are pretty close to being ready but there's still some fine tuning that could be done. It might be worth having @steeleb look over things at a high level if he has time, too. Mostly I'm just not super versed in package contributions so more eyes could be better. 

#### General
+ I like the way that the functions have been broken out and rewritten to complement one another 
+ Check for consistency in formatting and line breaks
+ There's commented out code that should get removed in `TADA_ViewATTAINS.R` L154, so you might double check elsewhere too

#### new_tada_functions.Rmd
I'm not 100% sure if this is intended as a long-term reference doc, so don't feel like you have to make these changes if it's not worth the work. Overall though I like having a document that goes through all the changes that you've made in one place!
+ A lot of the visual content of the knitted output is progress indicator info without code, which I think should kind of be flipped to make it more effective as a reference narrative. I ended up walking through the raw .Rmd instead of reading the HTML
+ Similarly to above, it might be more reader friendly to have code chunks  of single function runs preceeded by narrative text instead of having the explanatory text in comments like below:
```R
# This example doesn't work! TADA_dataset_3 has no data.
TADA_ViewATTAINS(data = TADA_dataset_3)
```

#### TADA_DataRetrieval_new.R
+ I think it would help if the documentation contained more guidance around the use of geographic and spatial arguments for this function. A few questions that came up for me as I used it:
   - Whether the user can or should provide **just** a shapefile (`sf =`) argument, just a state (`statecode =`) argument, etc. There's some language from TADA already about the AND/OR logic but the addition of `sf` made this less clear to me
   - What happens differently, if anything, when more than one geographic input is provided? i.e., does the shapefile subset within a state, county, huc, or site? When are these arguments redundant?
   - Does the `sf` data play into the WQP query at all? My original expectation was that it would, but then I saw that it did not (See my in-line comment about whether this is possible)

#### TADA_GetATTAINS.R
Your method with `return = T/F` seems appropriate to me!

#### Computation
When working through the changes to `TADA_DataRetrieval_new()` I started to wonder how computational efficiency would be affected by the addition of the `sf` argument. With that in mind, I think it would be worthwhile to test some of this out because reviewers or users would likely want to know if there were new limitations. Below is some code I tried to use to briefly test this out using variations of data size and different potential arguments getting at similar data requests:

```R
# Load only packages starting with "t":
library(TADA)
library(tictoc)
library(tidyverse)
library(tigris)

source("src/TADA_DataRetrieval_new.R")

azm_shp <- counties(state = "AZ") %>%
  filter(NAME == "Maricopa")


# One year of data: pH ----------------------------------------------------

# TADA currently
tic()
pH_1yr_orig <- TADA::TADA_DataRetrieval(startDate = "2019-01-01",
                                        endDate = "2020-01-01",
                                        characteristicName = "pH",
                                        statecode = "AZ",
                                        countycode = "Maricopa")
dim(pH_1yr_orig)
toc()
# 7.863 sec elapsed

# Now using our function with only a shapefile
tic()
pH_1yr_new <- TADA_DataRetrieval_new(startDate = "2019-01-01",
                                     endDate = "2020-01-01",
                                     characteristicName = "pH",
                                     sf = azm_shp)
dim(pH_1yr_new)
toc()
# 17.541 sec elapsed

# Now using our function with statecode and a shapefile
tic()
pH_1yr_new_plus <- TADA_DataRetrieval_new(startDate = "2019-01-01",
                                          endDate = "2020-01-01",
                                          characteristicName = "pH",
                                          statecode = "AZ",
                                          countycode = "Maricopa",
                                          sf = azm_shp)
dim(pH_1yr_new_plus)
toc()
# 13.637 sec elapsed


# Ten years of data: pH ---------------------------------------------------
# TADA currently
tic()
pH_10yr_orig <- TADA::TADA_DataRetrieval(startDate = "2010-01-01",
                                         endDate = "2020-01-01",
                                         characteristicName = "pH",
                                         statecode = "AZ",
                                         countycode = "Maricopa")
dim(pH_10yr_orig)
toc()
# 242.016 sec elapsed

# Now using our function with only a shapefile
tic()
pH_10yr_new <- TADA_DataRetrieval_new(startDate = "2010-01-01",
                                      endDate = "2020-01-01",
                                      characteristicName = "pH",
                                      sf = azm_shp)
dim(pH_10yr_new)
toc()
# Errored, unsure why

# Now using our function with statecode and a shapefile
tic()
pH_10yr_new_plus <- TADA_DataRetrieval_new(startDate = "2010-01-01",
                                           endDate = "2020-01-01",
                                           characteristicName = "pH",
                                           statecode = "AZ",
                                           countycode = "Maricopa",
                                           sf = azm_shp)
dim(pH_10yr_new_plus)
toc()
# 216.693 sec elapsed
```

I'm guessing that the compute time will vary by time of day for requests and could be affected in non-linear ways by our code changes. But a couple things that might be worth testing are:
+ Varying sizes of data requests
+ `sf` inputs using POLYGON vs MULTIPOLYGON vs. iterating over a list of single POLYGON objects
+ Providing a POLYGON object that has the same footprint as a state, county, etc. (I guess this assumes that the function allowed queries to WQP using spatial data)

I had more difficulty getting some things to run than I expected when testing speeds above, `pH_10yr_new` for example. In a previous attempt I also tried unsuccessfully to request an entire state's data using an sf boundary for the state. (I think the state was CO). So if there are things I'm doing wrong with the functions or incorrect assumptions that I made then those might cause others to stumble as well.

Let me know where you have questions or need clarifications!

steeleb commented:
Thanks for picking up my slack @mbrousil! I'll definitely take a look at this to see if I have anything to add shortly.

steeleb commented:
I'll be honest that I don't have much to add to @mbrousil's review! I, too, feel like this is foreign land (package development).

## new_tada_functions.Rmd

It looks like TADA tends to set their chunk options as follows to avoid progress bar ridiculousness in their vignettes:
```{r setup, include = FALSE}
knitr::opts_chunk$set(
  eval = FALSE
)
```
I second MB's comments about the `new_tada_functions.Rmd` document and rendered .html. I agree that showing code and adding the comments as suggested would create great vignettes, and shouldn't take a whole lot of work from here. 

## TADA_data_retrieval_new()

I suspect that the issue with HUCs and sf's being slow has something to do with some combination of dataRetrieval back end and NLDI entanglements (for the HUCS at least). I'll be honest that I don't know enough about either of these to diagnose next steps. In the short term, my suggestion would be to try the TADA_data_retrieval_new() function in a more.... hacky? ... way where you get the overlap of state/counties and pull that WQP data, then subset the resulting data by the sf or HUC. You mentioned this in our conversation, but testing the internals of that function to find out where the pain points are would probably help determine if this is an okay intermediary step. This may be an issue out of our control, but could be a great excuse to engage the USGS dataRetrieval team to help sleuth out what is causing the slowness if we can't track it down.


My suggestion in the near term would be to submit the ATTAINS functions, but wait on the addition of HUCs/sf to TADA_data_retrieval_new() until we know what is the cause of the slowness (and failing, in the case of MB's pH request). I think this also makes sense from a PR standpoint, providing a very tight group of functions that are related for EPA/TADA folks to review and creating vignettes specific to the ATTAINS functions. We could probably also document the slowness with Christina once we have an action plan for remediating it, to keep the conversation going with her. 

Thanks for being patient while I wrapped some other things up. Let me know if there is anything else I can do!

kathryn-willi committed:
Vignette for TADA PR

kathryn-willi committed:
New function to pull in ATTAINS data iteratively.

kathryn-willi committed:
Better roxygen + in-line comments,  added fetchATTAINS to make ATTAINS functions more concise.

kathryn-willi commented:
Hi y'all - based on some hurdles with the TADA_DataRetrieval_new(), I'm going to turn our discussions surrounding it into an issue so that we can deal with it later. Instead of trying to fix it, I've focused my efforts on finalizing  the other functions: `TADA_MakeSpatial()`, `TADA_GetATTAINS()`, and `TADA_ViewATTAINS()`. I also added a new function, `fetchATTAINS()` that is an intermediate step in pulling ATTAINS data that I don't think should be "exported" (ie @export) in the package itself just yet. I also added a vignette describing these functions. But! Based on Cristina's recommendation that we submit new code as a PR from a forked TADA repo, I'm shifting this work into the forked TADA repo I make in the ROSSyndicate org. Please expect a new, fresh PR from me there!
